From d22bd18490610652ab8ca6d8bda9ac773eb9c98f Mon Sep 17 00:00:00 2001
From: Stefano Gurrieri <stefano.gurrieri@roj.com>
Date: Thu, 16 Nov 2017 12:02:55 +0100
Subject: [PATCH] video-mode-settings

---
 include/configs/mx6smarc_common.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/mx6smarc_common.h b/include/configs/mx6smarc_common.h
index 0729194..06a4437 100644
--- a/include/configs/mx6smarc_common.h
+++ b/include/configs/mx6smarc_common.h
@@ -161,7 +161,7 @@
 		"fi\0" \
 	EMMC_ENV	  \
 	"smp=" SYS_NOSMP "\0"\
-	"mmcsetroot=setenv mmcroot /dev/mmcblk${mmcdevk}p${mmcrootpart} rootwait rw\0" \
+	"mmcsetroot=setenv mmcroot /dev/mmcblk${mmcdevk}p${mmcrootpart} rootwait rw video=mxcfb0:dev=ldb,800x600M@60,if=RGB24,fbpix=RGB32 video=mxcfb1:dev=hdmi,1920x1080M@60,if=RGB24,fbpix=RGB32\0" \
 	"mmcargs=run mmcsetroot;setenv bootargs console=${console},${baudrate} ${smp} " \
 		"root=${mmcroot}\0" \
 	"loadbootscript=" \
-- 
2.7.4

