From 17d655bc1c435a6e3e19aad984edf6ff7d3f3bfd Mon Sep 17 00:00:00 2001
From: Stefano Gurrieri <stefano.gurrieri@roj.com>
Date: Tue, 22 Oct 2019 15:57:51 +0200
Subject: [PATCH] spi-flash-support


diff --git a/arch/arm/boot/dts/imx6dl-enuc.dts b/arch/arm/boot/dts/imx6dl-enuc.dts
index 94076fd26..0c4b4a96d 100644
--- a/arch/arm/boot/dts/imx6dl-enuc.dts
+++ b/arch/arm/boot/dts/imx6dl-enuc.dts
@@ -67,6 +67,24 @@
 		};
 	};
 
+	/*SPI0*/
+	ecspi4 {
+		pinctrl_ecspi4_cs_1: ecspi4_cs_grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D29__GPIO3_IO29 0x100b1
+				MX6QDL_PAD_EIM_A25__GPIO5_IO02 0x100b1
+			>;
+		};
+
+		pinctrl_ecspi4_1: ecspi4grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D22__ECSPI4_MISO 0x100b1
+				MX6QDL_PAD_EIM_D28__ECSPI4_MOSI 0x100b1
+				MX6QDL_PAD_EIM_D21__ECSPI4_SCLK 0x100b1
+			>;
+		};
+	};
+
 };
 
 &regulators {
@@ -91,3 +109,19 @@
 			compatible = "brcm,bcm4329-fmac";
 		};	
 };
+
+&ecspi4 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "w25x20", "jedec,spi-nor";
+		spi-max-frequency = <20000000>;
+		reg = <0>;
+	};
+
+	flash@1 {
+		compatible = "w25x20", "jedec,spi-nor";
+		spi-max-frequency = <20000000>;
+		reg = <1>;
+	};
+};
diff --git a/arch/arm/boot/dts/imx6q-enuc.dts b/arch/arm/boot/dts/imx6q-enuc.dts
index 24f31cf2b..2d525d850 100644
--- a/arch/arm/boot/dts/imx6q-enuc.dts
+++ b/arch/arm/boot/dts/imx6q-enuc.dts
@@ -67,6 +67,24 @@
 		};
 	};
 
+	/*SPI0*/
+	ecspi4 {
+		pinctrl_ecspi4_cs_1: ecspi4_cs_grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D29__GPIO3_IO29 0x100b1
+				MX6QDL_PAD_EIM_A25__GPIO5_IO02 0x100b1
+			>;
+		};
+
+		pinctrl_ecspi4_1: ecspi4grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D22__ECSPI4_MISO 0x100b1
+				MX6QDL_PAD_EIM_D28__ECSPI4_MOSI 0x100b1
+				MX6QDL_PAD_EIM_D21__ECSPI4_SCLK 0x100b1
+			>;
+		};
+	};
+
 };
 
 &regulators {
@@ -91,3 +109,19 @@
 			compatible = "brcm,bcm4329-fmac";
 		};	
 };
+
+&ecspi4 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "w25x20", "jedec,spi-nor";
+		spi-max-frequency = <20000000>;
+		reg = <0>;
+	};
+
+	flash@1 {
+		compatible = "w25x20", "jedec,spi-nor";
+		spi-max-frequency = <20000000>;
+		reg = <1>;
+	};
+};
-- 
2.17.1

